Kernel/VirtIO/VirtIOConsole.cpp

```diff
@@ -9,6 +9,41 @@
 
 namespace Kernel {
 
+VirtIOConsole::VMRingBuffer::VMRingBuffer(const StringView region_name, size_t capacity)
+    : m_region(MM.allocate_contiguous_kernel_region(capacity, region_name, Region::Access::Read | Region::Access::Write))
+    , m_capacity_in_bytes(capacity)
+{
+}
+
+bool VirtIOConsole::VMRingBuffer::copy_data_in(const UserOrKernelBuffer& buffer, size_t length, PhysicalAddress& start_of_copied_data, size_t& bytes_copied)
+{
+    size_t start_of_free_area = (m_start_of_used + m_num_used_bytes) % m_capacity_in_bytes;
+    bytes_copied = min(m_capacity_in_bytes - m_num_used_bytes, min(m_capacity_in_bytes - start_of_free_area, length));
```

Comment:
The expected API for VMRingBuffer is that you wrap it in a loop that keeps calling `copy_data_in()` keeping track of how many bytes were written in total (realistically, this would be at most 2 calls, but this is a common pattern for write). This is done because after each copy, the caller probably wants to do something with the (ptr, size) pair that was returned.

We have to accept a non-full copy, otherwise what would we do in the case that the buffer can't fit all the data?
The implementation of `Process::do_write()` that is called by `sys$write` already does the thing where it repeatedly calls `VirtioConsole::write` until all the data is across, and it even handles blocking on the devices block condition, there is no point duplicating this logic here.